In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import math

## 读取数据

In [2]:
# 数据加载
df_train = pd.read_csv('./GiveMeSomeCredit/cs-training.csv')
df_train.head()

Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0           1                 1                              0.766127   45   
1           2                 0                              0.957151   40   
2           3                 0                              0.658180   38   
3           4                 0                              0.233810   30   
4           5                 0                              0.907239   49   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                     2   0.802982         9120.0   
1                                     0   0.121876         2600.0   
2                                     1   0.085113         3042.0   
3                                     0   0.036050         3300.0   
4                                     1   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                               13                        0   
1                                4                        0   
2                                2                        1   
3                                5                        0   
4                                7                        0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                             6                                     0   
1                             0                                     0   
2                             0                                     0   
3                             0                                     0   
4                             1                                     0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

## 数据探索和预处理

In [3]:
# Step 1
#去掉第一列
df_train = df_train.iloc[:,1:]

In [4]:
#缺失值情况
df_train.isnull().sum()

SeriousDlqin2yrs                            0
RevolvingUtilizationOfUnsecuredLines        0
age                                         0
NumberOfTime30-59DaysPastDueNotWorse        0
DebtRatio                                   0
MonthlyIncome                           29731
NumberOfOpenCreditLinesAndLoans             0
NumberOfTimes90DaysLate                     0
NumberRealEstateLoansOrLines                0
NumberOfTime60-89DaysPastDueNotWorse        0
NumberOfDependents                       3924
dtype: int64

In [5]:
# step 2
# 使用中位数的方式填充缺失值
df_train = df_train.fillna(df_train.median())
df_train.isnull().sum()

SeriousDlqin2yrs                        0
RevolvingUtilizationOfUnsecuredLines    0
age                                     0
NumberOfTime30-59DaysPastDueNotWorse    0
DebtRatio                               0
MonthlyIncome                           0
NumberOfOpenCreditLinesAndLoans         0
NumberOfTimes90DaysLate                 0
NumberRealEstateLoansOrLines            0
NumberOfTime60-89DaysPastDueNotWorse    0
NumberOfDependents                      0
dtype: int64

### 数据分箱

In [6]:
# step 3
# 将age字段进行分箱
age_bins = [-math.inf, 25, 40, 50, 60, 70, math.inf]
df_train['bin_age'] = pd.cut(df_train['age'],bins=age_bins)
df_train[['age','bin_age']]

age       bin_age
0        45  (40.0, 50.0]
1        40  (25.0, 40.0]
2        38  (25.0, 40.0]
3        30  (25.0, 40.0]
4        49  (40.0, 50.0]
...     ...           ...
149995   74   (70.0, inf]
149996   44  (40.0, 50.0]
149997   58  (50.0, 60.0]
149998   30  (25.0, 40.0]
149999   64  (60.0, 70.0]

[150000 rows x 2 columns]

In [7]:
#对于NumberOfDependents字段的分箱
dependents_bins = [-math.inf, 2, 4, 6, 8, 10, math.inf]
df_train['bin_NumberOfDependents'] = pd.cut(df_train['NumberOfDependents'],bins=dependents_bins)
df_train[['NumberOfDependents','bin_NumberOfDependents']]

NumberOfDependents bin_NumberOfDependents
0                      2.0            (-inf, 2.0]
1                      1.0            (-inf, 2.0]
2                      0.0            (-inf, 2.0]
3                      0.0            (-inf, 2.0]
4                      0.0            (-inf, 2.0]
...                    ...                    ...
149995                 0.0            (-inf, 2.0]
149996                 2.0            (-inf, 2.0]
149997                 0.0            (-inf, 2.0]
149998                 0.0            (-inf, 2.0]
149999                 0.0            (-inf, 2.0]

[150000 rows x 2 columns]

In [8]:
# 对于3种逾期次数，即NumberOfTime30-59DaysPastDueNotWorse，NumberOfTime60-89DaysPastDueNotWorse，NumberOfTimes90DaysLate，各分成10段
dpd_bins = [-math.inf,1,2,3,4,5,6,7,8,9,math.inf]
df_train['bin_NumberOfTime30-59DaysPastDueNotWorse'] = pd.cut(df_train['NumberOfTime30-59DaysPastDueNotWorse'],bins=dpd_bins)
df_train['bin_NumberOfTime60-89DaysPastDueNotWorse'] = pd.cut(df_train['NumberOfTime60-89DaysPastDueNotWorse'],bins=dpd_bins)
df_train['bin_NumberOfTimes90DaysLate'] = pd.cut(df_train['NumberOfTimes90DaysLate'],bins=dpd_bins)
df_train[['NumberOfTime30-59DaysPastDueNotWorse','bin_NumberOfTime30-59DaysPastDueNotWorse','NumberOfTime60-89DaysPastDueNotWorse',\
         'bin_NumberOfTime60-89DaysPastDueNotWorse','NumberOfTimes90DaysLate','bin_NumberOfTimes90DaysLate']]

NumberOfTime30-59DaysPastDueNotWorse  \
0                                          2   
1                                          0   
2                                          1   
3                                          0   
4                                          1   
...                                      ...   
149995                                     0   
149996                                     0   
149997                                     0   
149998                                     0   
149999                                     0   

       bin_NumberOfTime30-59DaysPastDueNotWorse  \
0                                    (1.0, 2.0]   
1                                   (-inf, 1.0]   
2                                   (-inf, 1.0]   
3                                   (-inf, 1.0]   
4                                   (-inf, 1.0]   
...                                         ...   
149995                              (-inf, 1.0]   
149996                              (-inf, 1.0]   
149997                              (-inf, 1.0]   
149998                              (-inf, 1.0]   
149999                              (-inf, 1.0]   

        NumberOfTime60-89DaysPastDueNotWorse  \
0                                          0   
1                                          0   
2                                          0   
3                                          0   
4                                          0   
...                                      ...   
149995                                     0   
149996                                     0   
149997                                     0   
149998                                     0   
149999                                     0   

       bin_NumberOfTime60-89DaysPastDueNotWorse  NumberOfTimes90DaysLate  \
0                                   (-inf, 1.0]                        0   
1                                   (-inf, 1.0]                        0   
2                                   (-inf, 1.0]                        1   
3                                   (-inf, 1.0]                        0   
4                                   (-inf, 1.0]                        0   
...                                         ...                      ...   
149995                              (-inf, 1.0]                        0   
149996                              (-inf, 1.0]                        0   
149997                              (-inf, 1.0]                        0   
149998                              (-inf, 1.0]                        0   
149999                              (-inf, 1.0]                        0   

       bin_NumberOfTimes90DaysLate  
0                      (-inf, 1.0]  
1                      (-inf, 1.0]  
2                      (-inf, 1.0]  
3                      (-inf, 1.0]  
4                      (-inf, 1.0]  
...                            ...  
149995                 (-inf, 1.0]  
149996                 (-inf, 1.0]  
149997                 (-inf, 1.0]  
149998                 (-inf, 1.0]  
149999                 (-inf, 1.0]  

[150000 rows x 6 columns]

In [9]:
# 对于RevolvingUtilizationOfUnsecuredLines, DebtRatio, MonthlyIncome, 
# NumberOfOpenCreditLinesAndLoans, NumberRealEstateLoansOrLines 分成5段
df_train['bin_RevolvingUtilizationOfUnsecuredLines'] = pd.qcut(df_train['RevolvingUtilizationOfUnsecuredLines'],q=5,duplicates='drop')
df_train['bin_DebtRatio'] = pd.qcut(df_train['DebtRatio'],q=5,duplicates='drop')
df_train['bin_MonthlyIncome'] = pd.qcut(df_train['MonthlyIncome'],q=5,duplicates='drop')
df_train['bin_NumberOfOpenCreditLinesAndLoans'] = pd.qcut(df_train['NumberOfOpenCreditLinesAndLoans'],q=5,duplicates='drop')
loans_bins = [-math.inf,0,1,2,3,math.inf]
df_train['bin_NumberRealEstateLoansOrLines'] = pd.cut(df_train['NumberRealEstateLoansOrLines'],bins=loans_bins)
df_train.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0                 1                              0.766127   45   
1                 0                              0.957151   40   
2                 0                              0.658180   38   
3                 0                              0.233810   30   
4                 0                              0.907239   49   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                     2   0.802982         9120.0   
1                                     0   0.121876         2600.0   
2                                     1   0.085113         3042.0   
3                                     0   0.036050         3300.0   
4                                     1   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                               13                        0   
1                                4                        0   
2                                2                        1   
3                                5                        0   
4                                7                        0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  ...  \
0                             6                                     0  ...   
1                             0                                     0  ...   
2                             0                                     0  ...   
3                             0                                     0  ...   
4                             1                                     0  ...   

        bin_age bin_NumberOfDependents  \
0  (40.0, 50.0]            (-inf, 2.0]   
1  (25.0, 40.0]            (-inf, 2.0]   
2  (25.0, 40.0]            (-inf, 2.0]   
3  (25.0, 40.0]            (-inf, 2.0]   
4  (40.0, 50.0]            (-inf, 2.0]   

  bin_NumberOfTime30-59DaysPastDueNotWorse  \
0                               (1.0, 2.0]   
1                              (-inf, 1.0]   
2                              (-inf, 1.0]   
3                              (-inf, 1.0]   
4                              (-inf, 1.0]   

  bin_NumberOfTime60-89DaysPastDueNotWorse bin_NumberOfTimes90DaysLate  \
0                              (-inf, 1.0]                 (-inf, 1.0]   
1                              (-inf, 1.0]                 (-inf, 1.0]   
2                              (-inf, 1.0]                 (-inf, 1.0]   
3                              (-inf, 1.0]                 (-inf, 1.0]   
4                              (-inf, 1.0]                 (-inf, 1.0]   

  bin_RevolvingUtilizationOfUnsecuredLines    bin_DebtRatio  \
0                         (0.699, 50708.0]     (0.468, 4.0]   
1                         (0.699, 50708.0]  (-0.001, 0.134]   
2                           (0.271, 0.699]  (-0.001, 0.134]   
3                          (0.0832, 0.271]  (-0.001, 0.134]   
4                         (0.699, 50708.0]  (-0.001, 0.134]   

     bin_MonthlyIncome bin_NumberOfOpenCreditLinesAndLoans  \
0  (8250.0, 3008750.0]                        (12.0, 58.0]   
1     (-0.001, 3400.0]                       (-0.001, 4.0]   
2     (-0.001, 3400.0]                       (-0.001, 4.0]   
3     (-0.001, 3400.0]                          (4.0, 6.0]   
4  (8250.0, 3008750.0]                          (6.0, 9.0]   

  bin_NumberRealEstateLoansOrLines  
0                       (3.0, inf]  
1                      (-inf, 0.0]  
2                      (-inf, 0.0]  
3                      (-inf, 0.0]  
4                       (0.0, 1.0]  

[5 rows x 21 columns]

In [10]:
df_train['bin_NumberRealEstateLoansOrLines'].value_counts()

(-inf, 0.0]    56188
(0.0, 1.0]     52338
(1.0, 2.0]     31522
(2.0, 3.0]      6300
(3.0, inf]      3652
Name: bin_NumberRealEstateLoansOrLines, dtype: int64

In [11]:
# 只保留分箱字段
bin_cols = [c for c in df_train.columns.values if c.startswith('bin_')]

In [12]:
# step 4
def IV_result(df,feature,target):
    lst = []
    cols = ['Variable','Value','All','Bad']
    for i in range(df[feature].nunique()):
        val = list(df[feature].unique())[i]
        lst.append([feature,val,df[df[feature]==val].count()[feature],\
                   df[(df[feature]==val) & (df[target]==1)].count()[feature]])

    data = pd.DataFrame(lst,columns=cols)
    data = data[data['Bad'] > 0]
    
    data['Share'] = data['All'] / data['All'].sum() 
    data['Bad Rate'] = data['Bad'] / data['All'] 
    data['Margin Bad'] = data['Bad'] / data['Bad'].sum()
    data['Margin Good'] = (data['All'] - data['Bad']) / (data['All'].sum() - data['Bad'].sum())
    data['woe'] = np.log(data['Margin Bad'] / data['Margin Good'])
    data['iv'] = (data['woe'] * (data['Margin Bad'] - data['Margin Good'])).sum()
    data.sort_values(by = ['Variable','Value'])
    return data['iv'].values[0]

In [13]:
# 衡量自变量的预测能力，检测IV>0.1的
for f in bin_cols:
    if IV_result(df_train,f,'SeriousDlqin2yrs') >= 0.1:
        print(f,IV_result(df_train,f,'SeriousDlqin2yrs'))

bin_age 0.24041120302785982
bin_NumberOfTime30-59DaysPastDueNotWorse 0.492444774570198
bin_NumberOfTime60-89DaysPastDueNotWorse 0.2665587583516951
bin_NumberOfTimes90DaysLate 0.49160685733515563
bin_RevolvingUtilizationOfUnsecuredLines 1.0596188771423887


In [14]:
# step 5
# 计算woe
def WOE_result(df,features,target):
    df_new = df.copy()
    for f in features:
        df_woe = df_new.groupby(f).agg({target:['sum','count']})
        df_woe.columns = list(map(''.join,df_woe.columns.values))
        df_woe = df_woe.reset_index()
        df_woe = df_woe.rename(columns={target+'sum':'bad',target+'count':'all'})

        df_woe['good'] = df_woe['all'] - df_woe['bad']
        df_woe['margin bad'] = df_woe['bad'] / df_woe['bad'].sum()
        df_woe['margin good'] = df_woe['good'] / df_woe['good'].sum()
        df_woe['woe'] = np.log1p(df_woe['margin bad'] / df_woe['margin good'])
        df_woe.columns = [c if c==f else c+'_'+f for c in list(df_woe.columns.values)]
        df_new = df_new.merge(df_woe,on=f,how='left')
    return df_new

In [15]:
# 计算这些特征的woe
df_woe = WOE_result(df_train,bin_cols,'SeriousDlqin2yrs')

In [16]:
# 得到WOE的规则
feature_columns = ['age','NumberOfTime30-59DaysPastDueNotWorse',\
                'NumberOfTime60-89DaysPastDueNotWorse','NumberOfTimes90DaysLate',\
               'RevolvingUtilizationOfUnsecuredLines']

df_bin_to_woe = pd.DataFrame(columns=['features','bin','woe'])
for f in feature_columns:
    b = 'bin_' + f
    w = 'woe_bin_' + f
    df = df_woe[[w,b]].drop_duplicates()
    df.columns = ['woe','bin']
    df['features'] = f
    df_bin_to_woe = pd.concat([df_bin_to_woe,df])

In [17]:
df_bin_to_woe

features               bin       woe
0                                       age      (40.0, 50.0]  0.813822
1                                       age      (25.0, 40.0]  0.955231
5                                       age       (70.0, inf]  0.279404
6                                       age      (50.0, 60.0]  0.651655
15                                      age      (60.0, 70.0]  0.406848
19                                      age      (-inf, 25.0]  1.013134
0      NumberOfTime30-59DaysPastDueNotWorse        (1.0, 2.0]  1.797837
1      NumberOfTime30-59DaysPastDueNotWorse       (-inf, 1.0]  0.572521
13     NumberOfTime30-59DaysPastDueNotWorse        (2.0, 3.0]  2.151185
183    NumberOfTime30-59DaysPastDueNotWorse        (3.0, 4.0]  2.429111
191    NumberOfTime30-59DaysPastDueNotWorse        (4.0, 5.0]  2.520613
251    NumberOfTime30-59DaysPastDueNotWorse        (6.0, 7.0]  2.774776
423    NumberOfTime30-59DaysPastDueNotWorse        (9.0, inf]  2.902860
1052   NumberOfTime30-59DaysPastDueNotWorse        (5.0, 6.0]  2.812612
6909   NumberOfTime30-59DaysPastDueNotWorse        (7.0, 8.0]  2.024184
10822  NumberOfTime30-59DaysPastDueNotWorse        (8.0, 9.0]  2.077007
0      NumberOfTime60-89DaysPastDueNotWorse       (-inf, 1.0]  0.645352
186    NumberOfTime60-89DaysPastDueNotWorse        (1.0, 2.0]  2.712133
423    NumberOfTime60-89DaysPastDueNotWorse        (4.0, 5.0]  3.159234
1146   NumberOfTime60-89DaysPastDueNotWorse        (2.0, 3.0]  2.955438
1733   NumberOfTime60-89DaysPastDueNotWorse        (9.0, inf]  2.886833
2406   NumberOfTime60-89DaysPastDueNotWorse        (3.0, 4.0]  3.164917
6664   NumberOfTime60-89DaysPastDueNotWorse        (5.0, 6.0]  3.758483
16642  NumberOfTime60-89DaysPastDueNotWorse        (6.0, 7.0]  2.915139
23964  NumberOfTime60-89DaysPastDueNotWorse        (7.0, 8.0]  2.705454
68976  NumberOfTime60-89DaysPastDueNotWorse        (8.0, 9.0]  0.000000
0                   NumberOfTimes90DaysLate       (-inf, 1.0]  0.608707
13                  NumberOfTimes90DaysLate        (2.0, 3.0]  2.998746
186                 NumberOfTimes90DaysLate        (1.0, 2.0]  2.701853
1298                NumberOfTimes90DaysLate        (4.0, 5.0]  3.224503
1713                NumberOfTimes90DaysLate        (3.0, 4.0]  3.379582
1733                NumberOfTimes90DaysLate        (9.0, inf]  2.878935
2910                NumberOfTimes90DaysLate        (8.0, 9.0]  3.691154
3400                NumberOfTimes90DaysLate        (5.0, 6.0]  3.088387
3929                NumberOfTimes90DaysLate        (6.0, 7.0]  4.140397
5684                NumberOfTimes90DaysLate        (7.0, 8.0]  3.580814
0      RevolvingUtilizationOfUnsecuredLines  (0.699, 50708.0]  1.495914
2      RevolvingUtilizationOfUnsecuredLines    (0.271, 0.699]  0.720083
3      RevolvingUtilizationOfUnsecuredLines   (0.0832, 0.271]  0.350952
11     RevolvingUtilizationOfUnsecuredLines  (-0.001, 0.0192]  0.243890
14     RevolvingUtilizationOfUnsecuredLines  (0.0192, 0.0832]  0.211221

In [18]:
woe_cols = [c for c in list(df_woe.columns.values) if 'woe' in c]

In [19]:
df_woe[woe_cols]

woe_bin_age  woe_bin_NumberOfDependents  \
0          0.813822                     0.67296   
1          0.955231                     0.67296   
2          0.955231                     0.67296   
3          0.955231                     0.67296   
4          0.813822                     0.67296   
...             ...                         ...   
149995     0.279404                     0.67296   
149996     0.813822                     0.67296   
149997     0.651655                     0.67296   
149998     0.955231                     0.67296   
149999     0.406848                     0.67296   

        woe_bin_NumberOfTime30-59DaysPastDueNotWorse  \
0                                           1.797837   
1                                           0.572521   
2                                           0.572521   
3                                           0.572521   
4                                           0.572521   
...                                              ...   
149995                                      0.572521   
149996                                      0.572521   
149997                                      0.572521   
149998                                      0.572521   
149999                                      0.572521   

        woe_bin_NumberOfTime60-89DaysPastDueNotWorse  \
0                                           0.645352   
1                                           0.645352   
2                                           0.645352   
3                                           0.645352   
4                                           0.645352   
...                                              ...   
149995                                      0.645352   
149996                                      0.645352   
149997                                      0.645352   
149998                                      0.645352   
149999                                      0.645352   

        woe_bin_NumberOfTimes90DaysLate  \
0                              0.608707   
1                              0.608707   
2                              0.608707   
3                              0.608707   
4                              0.608707   
...                                 ...   
149995                         0.608707   
149996                         0.608707   
149997                         0.608707   
149998                         0.608707   
149999                         0.608707   

        woe_bin_RevolvingUtilizationOfUnsecuredLines  woe_bin_DebtRatio  \
0                                           1.495914           0.928274   
1                                           1.495914           0.645506   
2                                           0.720083           0.645506   
3                                           0.350952           0.645506   
4                                           1.495914           0.645506   
...                                              ...                ...   
149995                                      0.211221           0.613576   
149996                                      0.720083           0.928274   
149997                                      0.350952           0.597328   
149998                                      0.243890           0.645506   
149999                                      1.495914           0.613576   

        woe_bin_MonthlyIncome  woe_bin_NumberOfOpenCreditLinesAndLoans  \
0                    0.516960                                 0.691873   
1                    0.882076                                 0.882845   
2                    0.882076                                 0.882845   
3                    0.882076                                 0.630962   
4                    0.516960                                 0.588475   
...                       ...                                      ...   
149995               0.882076                                 0.882845   
149996               0.643114          

## 建模
### 逻辑回归

In [20]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,roc_auc_score

In [21]:
# 逻辑回归模型
x_train,x_test,y_train,y_test = train_test_split(df_woe[woe_cols],df_woe['SeriousDlqin2yrs'],test_size=0.2,random_state = 2021)

In [23]:
model = LogisticRegression(random_state=2021).fit(x_train,y_train)
y_pred = model.predict(x_test)
print('acc:',accuracy_score(y_pred,y_test))
print('auc:', roc_auc_score(y_pred,y_test))

acc: 0.9361
auc: 0.782286489205088


In [27]:
feature_cols = ['RevolvingUtilizationOfUnsecuredLines', 'age',
       'NumberOfTime30-59DaysPastDueNotWorse', 'DebtRatio', 'MonthlyIncome',
       'NumberOfOpenCreditLinesAndLoans', 'NumberOfTimes90DaysLate',
       'NumberRealEstateLoansOrLines', 'NumberOfTime60-89DaysPastDueNotWorse',
       'NumberOfDependents']

In [62]:
A = 650 # 基准分
B = 72.13 # PDO /ln(2)
def generate_scoredcard(model_colef, df_rule, features, B):
    coef = model_colef[0]
    lst = []
    cols = ['Variable','Binning','Score']
    for i in range(len(features)):
        feature = features[i]
        # 筛选该f的WOE规则
        df = df_rule[df_rule['features'] == feature]
        for index, row in df.iterrows():
            lst.append([feature, row['bin'], int(round(-B * coef[i] * row['woe']))])
    data = pd.DataFrame(lst, columns=cols)
    return data

In [68]:
score_card = generate_scoredcard(model.coef_, df_bin_to_woe, feature_cols, B)

In [83]:
def str_to_int(x):
    if x == '-inf':
        return -999999
    elif x == 'inf':
        return 999999
    return float(x)
    
# 将value映射到bin
def map_value_to_bin(feature_value, feature_to_bin):
    for index, row in feature_to_bin.iterrows():
        bins = str(row['Binning'])
        binnings = bins[1:-1].split(',')
        in_range = True
        if feature_value <= str_to_int(binnings[0]):
            in_range = False
        if feature_value > str_to_int(binnings[1]):
            in_range = False
        if in_range:
            return row['Binning']
    return null
    
def map_to_score(df, score_card):
    scored_columns = list(score_card['Variable'].unique())
    score = 0
    for col in scored_columns:
        feature_to_bin = score_card[score_card['Variable']==col]
        feature_value = df[col]
        selected_bin = map_value_to_bin(feature_value, feature_to_bin)
        selected_score = feature_to_bin[feature_to_bin['Binning'] == selected_bin]['Score'].iloc[0]
        score += selected_score
    return score

# 按照评分卡规则计算df的分数        
def cal_score_with_card(df, score_card, A):
    df['score'] = df.apply(map_to_score, args=(score_card, ), axis=1)
    df['score'] = df['score'] + A
    df['score'] = df['score'] - B * model.intercept_[0]
    df['score'] = df['score'].astype(int)
    return df

In [84]:
# 随即筛选好人
good_sample = df_train[df_train['SeriousDlqin2yrs'] == 0].sample(5)
good_sample = good_sample[feature_cols]
# 计算分数
cal_score_with_card(good_sample, score_card, 650)

RevolvingUtilizationOfUnsecuredLines  age  \
91675                               1.000000   51   
110252                              0.409753   65   
93352                               0.023823   63   
20653                               0.875422   51   
104679                              0.192901   33   

        NumberOfTime30-59DaysPastDueNotWorse    DebtRatio  MonthlyIncome  \
91675                                      0  2428.000000         5400.0   
110252                                     0     0.093245         8600.0   
93352                                      0     0.099800        10500.0   
20653                                      1     0.260091        30000.0   
104679                                     0     0.438113         2140.0   

        NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
91675                                 3                        0   
110252                                7                        0   
93352                                10                        0   
20653                                10                        0   
104679                                4                        0   

        NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
91675                              1                                     0   
110252                             0                                     0   
93352                              1                                     0   
20653                              3                                     0   
104679                             1                                     0   

        NumberOfDependents  score  
91675                  1.0   1043  
110252                 1.0   1111  
93352                  0.0   1146  
20653                  0.0   1043  
104679                 0.0   1105

In [85]:
# 随即筛选坏人
bad_sample = df_train[df_train['SeriousDlqin2yrs'] == 1].sample(5)
bad_sample = bad_sample[feature_cols]
# 计算分数
cal_score_with_card(bad_sample, score_card, 650)

RevolvingUtilizationOfUnsecuredLines  age  \
116179                              0.066924   50   
119410                              0.760592   62   
133004                              1.210526   38   
60851                               0.383482   50   
117978                              0.465116   47   

        NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
116179                                     0   0.460769         3300.0   
119410                                     0   1.077968         6848.0   
133004                                     1   0.012568         3500.0   
60851                                      0   0.525299         9881.0   
117978                                     5   0.245710         7166.0   

        NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
116179                                5                        0   
119410                               16                        0   
133004                                2                        7   
60851                                12                        0   
117978                                4                        1   

        NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
116179                             2                                     0   
119410                             2                                     0   
133004                             0                                     1   
60851                              1                                     0   
117978                             1                                     0   

        NumberOfDependents  score  
116179                 1.0   1123  
119410                 0.0   1057  
133004                 0.0    651  
60851                  1.0   1088  
117978                 1.0    964

我自己算的好人坏人的得分差比较小

In [86]:
score_card

Variable           Binning  Score
0   RevolvingUtilizationOfUnsecuredLines  (0.699, 50708.0]   -103
1   RevolvingUtilizationOfUnsecuredLines    (0.271, 0.699]    -49
2   RevolvingUtilizationOfUnsecuredLines   (0.0832, 0.271]    -24
3   RevolvingUtilizationOfUnsecuredLines  (-0.001, 0.0192]    -17
4   RevolvingUtilizationOfUnsecuredLines  (0.0192, 0.0832]    -14
5                                    age      (40.0, 50.0]    -47
6                                    age      (25.0, 40.0]    -55
7                                    age       (70.0, inf]    -16
8                                    age      (50.0, 60.0]    -38
9                                    age      (60.0, 70.0]    -24
10                                   age      (-inf, 25.0]    -59
11  NumberOfTime30-59DaysPastDueNotWorse        (1.0, 2.0]   -114
12  NumberOfTime30-59DaysPastDueNotWorse       (-inf, 1.0]    -36
13  NumberOfTime30-59DaysPastDueNotWorse        (2.0, 3.0]   -137
14  NumberOfTime30-59DaysPastDueNotWorse        (3.0, 4.0]   -154
15  NumberOfTime30-59DaysPastDueNotWorse        (4.0, 5.0]   -160
16  NumberOfTime30-59DaysPastDueNotWorse        (6.0, 7.0]   -176
17  NumberOfTime30-59DaysPastDueNotWorse        (9.0, inf]   -184
18  NumberOfTime30-59DaysPastDueNotWorse        (5.0, 6.0]   -179
19  NumberOfTime30-59DaysPastDueNotWorse        (7.0, 8.0]   -129
20  NumberOfTime30-59DaysPastDueNotWorse        (8.0, 9.0]   -132
21               NumberOfTimes90DaysLate       (-inf, 1.0]    -65
22               NumberOfTimes90DaysLate        (2.0, 3.0]   -319
23               NumberOfTimes90DaysLate        (1.0, 2.0]   -287
24               NumberOfTimes90DaysLate        (4.0, 5.0]   -343
25               NumberOfTimes90DaysLate        (3.0, 4.0]   -359
26               NumberOfTimes90DaysLate        (9.0, inf]   -306
27               NumberOfTimes90DaysLate        (8.0, 9.0]   -392
28               NumberOfTimes90DaysLate        (5.0, 6.0]   -328
29               NumberOfTimes90DaysLate        (6.0, 7.0]   -440
30               NumberOfTimes90DaysLate        (7.0, 8.0]   -380
31  NumberOfTime60-89DaysPastDueNotWorse       (-inf, 1.0]      1
32  NumberOfTime60-89DaysPastDueNotWorse        (1.0, 2.0]      6
33  NumberOfTime60-89DaysPastDueNotWorse        (4.0, 5.0]      7
34  NumberOfTime60-89DaysPastDueNotWorse        (2.0, 3.0]      6
35  NumberOfTime60-89DaysPastDueNotWorse        (9.0, inf]      6
36  NumberOfTime60-89DaysPastDueNotWorse        (3.0, 4.0]      7
37  NumberOfTime60-89DaysPastDueNotWorse        (5.0, 6.0]      8
38  NumberOfTime60-89DaysPastDueNotWorse        (6.0, 7.0]      6
39  NumberOfTime60-89DaysPastDueNotWorse        (7.0, 8.0]      6
40  NumberOfTime60-89DaysPastDueNotWorse        (8.0, 9.0]      0

**评分卡分析**

1、借款人当时的年龄 age
25岁以下以及25-40岁这个年龄段的人存在较大借贷风险；  
40-50这个年龄段相比也有较大借贷风险，但是比前面的略有降低；  
50-60这个年龄段相比也有较大借贷风险，但是比40-50也略有降低；  
60-70和70以上的借贷风险一般。

2、两年内35-59天逾期次数 NumberOfTime30-59DaysPastDueNotWorse
0和1这个阶段相对来说不是很危险；  
2，3，8这三个阶段有些危险；  
4，5，6，7，9以及以上都非常危险；


3、两年内90天或高于90天逾期的次数 NumberOfTimes90DaysLate
0和1次相对来说不是很危险；  
剩下基本都在一个数量级上，都非常危险；


4、包括了开放式信贷和贷款数量、不动产贷款或额度数量。 RevolvingUtilizationOfUnsecuredLines
(-0.001, 0.0192],(0.0192, 0.0832],(0.0832, 0.271]这三个范围的用户，贷款的数量轻微高过自己所有的资产量，因此具备良好的还款能力；    
(0.271, 0.699]	分数比值在这个范围的用户，贷款的数量略高过自己所有的资产量，因此具备一般的还款能力；  
(0.699, 50708.0] 分数比值在这个范围的用户，贷款的数量远高过自己所有的资产量，因此具备较差的还款能力。